File: packages/core/test/differential/python-harness.ts
import { execFileSync } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const harnessPath = join(__dirname, "python_harness.py");

/**
 * The Python interpreter used to run python_harness.py. Configurable (M4, Codex M3-review
 * follow-up item 3) so a machine with the private reference implementation installed under
 * a non-default interpreter (a venv, a specific pyenv version, etc.) can still run these
 * tests -- this repo itself never assumes a particular install location.
 */
const PYTHON_BIN = process.env.LANE_DIFFERENTIAL_PYTHON_BIN ?? "python3";

/**
 * Calls the real installed Python reference implementation package (v0.7.8) via
 * python_harness.py and returns its result. Throws if the Python side raised or the
 * harness rejected the function name.
 */
export function callPython<T = unknown>(fn: string, args: unknown[]): T {
  const stdout = execFileSync(PYTHON_BIN, [harnessPath], {
    input: JSON.stringify({ fn, args }),
    encoding: "utf-8",
  });
  const parsed = JSON.parse(stdout) as { result?: T; error?: string };
  if (parsed.error) {
    throw new Error(`python_harness.py: ${parsed.error}`);
  }
  return parsed.result as T;
}

let warnedAboutSkip = false;

/**
 * Whether the differential test suite can actually run right now (M4): explicitly
 * disabled via LANE_SKIP_DIFFERENTIAL_TESTS, or the private Python reference
 * implementation isn't importable at PYTHON_BIN. The reference implementation this repo
 * was ported from is not published anywhere the public can install it, so a fresh clone
 * (or CI without special setup) has no way to satisfy this -- differential tests skip
 * gracefully in that case rather than failing the whole suite (same
 * describe/describe.skip convention this repo already uses for real-subprocess tests
 * against agent-cost, e.g. packages/adapters/test/telemetry-agent-cost.test.ts).
 *
 * Codex M4 review, should-3: a skip is silent by default (vitest's own "N skipped" summary
 * line is easy to miss in CI output that also expects some tests to be skipped by design,
 * e.g. the real-subprocess adapter tests). Each call that resolves to "unavailable" also
 * prints a `::warning::` line -- GitHub Actions renders this as an actual annotation on
 * the job, not just scrollback text -- once per module load (each of the 4 differential
 * test files calls this once at module scope), so the parity suite being skipped is
 * something a maintainer would actually notice.
 */
export function isPythonReferenceAvailable(): boolean {
  const disabled = Boolean(process.env.LANE_SKIP_DIFFERENTIAL_TESTS);
  const available =
    !disabled &&
    (() => {
      try {
        execFileSync(PYTHON_BIN, ["-c", "import qureo_lane"], { stdio: "ignore" });
        return true;
      } catch {
        return false;
      }
    })();
  if (!available && !warnedAboutSkip) {
    warnedAboutSkip = true;
    console.warn(
      "::warning::differential parity suite skipped — python reference implementation unavailable " +
        "(set LANE_DIFFERENTIAL_PYTHON_BIN to an interpreter that has it installed, or LANE_SKIP_DIFFERENTIAL_TESTS was set explicitly)",
    );
  }
  return available;
}

File: packages/core/test/differential/python_harness.py
"""Differential test harness: calls the real, installed Python reference implementation
package (v0.9.0) so packages/core's TS ports can be checked byte-for-byte against the
Python implementation they were ported from (design.md §9 checkpoint 3 / §10).

This is a private reference implementation this repo was ported from, not published
anywhere the public can install it -- see docs/design.md §9/§10 and the README's
"differential tests" note for what this means for a fresh clone (they skip gracefully
when the package below isn't importable, per LANE_SKIP_DIFFERENTIAL_TESTS).

Usage: `echo '{"fn": "...", "args": [...]}' | python3 python_harness.py`
Prints `{"result": <value>}` or `{"error": "<message>"}` to stdout as JSON.

Only a fixed allow-list of pure reference-implementation functions is exposed -- this
harness is a test fixture, not a general RPC bridge into the Python package.
"""

import json
import sys

# The one place in this codebase that must reference the actual private package name --
# there is no way to test parity against it without importing it by its real name.
from qureo_lane import orchestrator as o
from qureo_lane import validate as v


def _compute_ledger_entry_id(lane_id, phase, source, pricing_version):
    return o.compute_ledger_entry_id(lane_id, phase, source, pricing_version)


def _derive_confidence(source, scope):
    return o.derive_confidence(source, scope)


def _classify_data_state(import_exit_ok, had_events, total_tokens):
    return o.classify_data_state(import_exit_ok, had_events, total_tokens)


def _derive_included_in_kpi(entry, ledger):
    return o.derive_included_in_kpi(entry, ledger)


def _recompute_included_in_kpi(ledger):
    return o.recompute_included_in_kpi(ledger)


def _is_forward_transition(current, target):
    return o.is_forward_transition(current, target)


def _apply_done_overlay(state, overlay):
    return o.apply_done_overlay(state, overlay)


def _validate_no_personal_dimensions(payload):
    return o.validate_no_personal_dimensions(payload)


def _derive_cost_credits(cost_usd, source):
    return o.derive_cost_credits(cost_usd, source)


def _normalize_criterion(text):
    return v.normalize_criterion(text)


FUNCTIONS = {
    "compute_ledger_entry_id": _compute_ledger_entry_id,
    "derive_confidence": _derive_confidence,
    "classify_data_state": _classify_data_state,
    "derive_included_in_kpi": _derive_included_in_kpi,
    "recompute_included_in_kpi": _recompute_included_in_kpi,
    "is_forward_transition": _is_forward_transition,
    "apply_done_overlay": _apply_done_overlay,
    "validate_no_personal_dimensions": _validate_no_personal_dimensions,
    "derive_cost_credits": _derive_cost_credits,
    "normalize_criterion": _normalize_criterion,
}


def main():
    request = json.loads(sys.stdin.read())
    fn_name = request["fn"]
    args = request.get("args", [])
    if fn_name not in FUNCTIONS:
        print(json.dumps({"error": f"unknown fn: {fn_name}"}))
        return
    try:
        result = FUNCTIONS[fn_name](*args)
        print(json.dumps({"result": result}))
    except Exception as exc:  # noqa: BLE001 - test harness, surface any failure as JSON
        print(json.dumps({"error": f"{type(exc).__name__}: {exc}"}))


if __name__ == "__main__":
    main()
